src/nvim/api/mark_extended.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Same function but names are not limited to one char
+ * There is also no ex_command, just viml functions
+ */
+
+#include "nvim/vim.h"
+#include "nvim/globals.h"      // FOR_ALL_BUFFERS
+#include "nvim/mark.h"         // SET_FMARK
+#include "nvim/memory.h"
+#include "nvim/map.h"          // pmap ...
+#include "nvim/lib/kbtree.h"   // kbitr ...
+#include "nvim/api/mark_extended.h"
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "api/extmark.c.generated.h"
+#endif
+
+#define FOR_ALL_EXTMARKS(buf) \
+  kbitr_t itr; \
+  ExtendedMark *extmark; \
+  for (;kb_itr_valid(&itr); kb_itr_next(ExtendedMark, buf->b_extmarks_tree, &itr)){
+
+/* Create or update an extmark, */
+int extmark_set(buf_T *buf, cstr_t *name, pos_T *pos)
```

Comment:
I actually have no idea why `*Map` interface exists, khash.h is already fine and it does not require you to query hash twice to do “if KEY is in hash return VAL, otherwise put (KEY, VAL) to hash” . Also one `map_new` is *two* small `malloc`s alone, while my small changes to khash.h allowed to put hash on stack (or e.g. directly as a part of some structure, not as a pointer).

My code never uses map.h.